### ISSUE TITLE
Support migration of existing Jira Cloud connection to use OAuth approach

### ISSUE THREAD

**Issue by mickmister** [mattermost/mattermost-plugin-jira#971]
In order to migrate an existing MM instance currently using Jira Cloud to use the OAuth approach, we need to add a way to simultaneously support both instance types from the same Jira URL. This is so webhooks/subscriptions etc. can continue to function properly, and existing connected users can continue to interact with the plugin, as the user base is in the process of migrating connections to OAuth.

_Issue created from a [Mattermost message](https://community.mattermost.com/_redirect/pl/qpjkoimuhfn4frqf9ky5eticsw) *by @mickmister*._

**Comment by mickmister** [comment]
I'm thinking this should not block the 4.0.0 release. Giving the current state of the code to a new customer is a good idea, but not to an existing system that wants to migrate over. Since the admin would need to explicitly run `/jira instance uninstall` to get them into a bad state, I think this can wait for next release. Though this is a particularly important thing to follow up on.

**Comment by raghavaggarwal2308** [comment]
@mickmister I am a bit confused about what exactly are you expecting here:
1. If you are asking that the user's old channel subscriptions still work after switching to OAuth, then I think that will still work with the current code as the URL on which Jira sends the webhook is the same for both. So, we will continue to get the webhook notification from Jira even after switching to OAuth.
2. If you are asking that the user not have to uninstall the JWT instance to switch to the OAuth instance, then also I think it with work with the current code. I tried to install the instance with both JWT and Oauth and when I uninstalled the OAuth one, the other one was uninstalled too. I think currently we are treating them as a single instance as the URL is the same.

Can you please explain a bit more about the issue here?

**Comment by mickmister** [comment]
> If you are asking that the user's old channel subscriptions still work after switching to OAuth, then I think that will still work with the current code as the URL on which Jira sends the webhook is the same for both. So, we will continue to get the webhook notification from Jira even after switching to OAuth.

@raghavaggarwal2308 Something that comes to mind here, is are channel subscriptions still active when you uninstall and reinstall a Jira Cloud instance?

>  I tried to install the instance with both JWT and Oauth and when I uninstalled the OAuth one, the other one was uninstalled too.

Are you saying that both can be installed at the same time already?

The main issue I'm thinking of is:

- I'm a user connected to JWT instance
- My admin wants to migrate to OAuth
- Admin uninstalls JWT, and installs OAuth
- I can't interact with Jira until I reconnect my account. Can't receive notifications etc.
    - I'm not sure if that's how it would work, but bringing it up as we may need to have a migration plan if that's the case

**Comment by raghavaggarwal2308** [comment]
@mickmister 
>@raghavaggarwal2308 Something that comes to mind here, is are channel subscriptions still active when you uninstall and reinstall a Jira Cloud instance?

Yes, the subscriptions are still valid.

>Are you saying that both can be installed at the same time already?

Not at the same time but the OAuth instance overrides the JWT instance if we install a OAuth instance without uninstalling the JWT instance. The reason for this is a check added in the code for installing OAuth instance.

![image](https://github.com/mattermost/mattermost-plugin-jira/assets/55234496/b06684f0-8786-4ad1-90e5-0b35b5b6de00)
Because of this condition, the code ignores that an instance with the same URL existed before and continues with the setup.

>The main issue I'm thinking of is:
>
>I'm a user connected to JWT instance
My admin wants to migrate to OAuth
Admin uninstalls JWT, and installs OAuth
I can't interact with Jira until I reconnect my account. Can't receive notifications etc.
I'm not sure if that's how it would work, but bringing it up as we may need to have a migration plan if that's the case

What do you mean by `migration plan` here? Are you suggesting that when an admin installs the OAuth instance, all the users who were connected with JWT earlier should be auto-connected to OAuth now?

**Comment by mickmister** [comment]
> Are you suggesting that when an admin installs the OAuth instance, all the users who were connected with JWT earlier should be auto-connected to OAuth now?

@raghavaggarwal2308 No I don't think that's possible. By "plan" I mean having a clear path forward for what the MM admin should do when moving off of the JWT app and onto the OAuth version.

Currently there are working deployments of the Jira Cloud JWT app in production (including the MM community server). It works correctly for all cases except for creating new user connections. If we were to run `/jira instance install cloud-oauth ...` on an existing deployment, what are the implications of this?

- Can I still receive DM notifications from the Jira bot, before reconnecting my account via OAuth?
- Can I still create issues within the MM UI, before reconnecting my account via OAuth?
- Do comment subscriptions still work that rely on the JWT "bot" account approach?

Essentially, we want the least amount of impact on existing functionality that users depend on, during the process of switching to the OAuth instance type. If possible, we should be able to support both instance types simultaneously, so that there is no disruption in working functionality. What are your thoughts on this?

**Comment by mickmister** [comment]
@raghavaggarwal2308 @Kshitij-Katiyar Note that this issue is being escalated due to issues occurring on the community server with the Jira plugin. We're trying to diagnose an issue with subscriptions, and some sysadmins are unable to help since they can't connect their Jira accounts. I would link to the thread on the discussion but it's in a private team.

What are your thoughts on my previous comment?

**Comment by raghavaggarwal2308** [comment]
@mickmister The best solution we could think of here is:
1. We update the enterprise check for this particular situation to consider instances with the same URL as a single instance but storing them separately in the KV store.
2. We need to update the key for storing the oauth instance something like instaceID_oauth in the KV store because if we don't, it will override the JWT instance. *Note:* This might cause some issues for users who have already installed oauth using the 4.0.0 release as they will already have an oauth instance stored with a different key in the KV store.

The code will keep using the JWT instance until the user has connected to the oauth instance.

**Comment by mickmister** [comment]
> This might cause some issues for users who have already installed oauth using the 4.0.0 release as they will already have an oauth instance stored with a different key in the KV store.

We definitely want to avoid that breaking change

> The code will keep using the JWT instance until the user has connected to the oauth instance.

I'm not sure I understand this. The plugin will continue to use the JWT connected token? So if I understand correctly, the user will be able to continue using their account. This would also take care of the subscription situation where need to fetch the comment body using the commenter's token.

@raghavaggarwal2308 @Kshitij-Katiyar

**Comment by raghavaggarwal2308** [comment]
> I'm not sure I understand this. The plugin will continue to use the JWT connected token? So if I understand correctly, the user will be able to continue using their account. This would also take care of the subscription situation where need to fetch the comment body using the commenter's token.

@mickmister Yes, we will be using the JWT connected token. Please share your views if we are not using this approach.

**Comment by raghavaggarwal2308** [comment]
Link to the community thread regarding this issue:
https://community.mattermost.com/core/pl/rj5ewcdo4ig73p8mk3ekkbbx4y
